### ISSUE TITLE
hasattr() + truthiness guard incorrectly infers ~AlwaysFalsy for attribute access

### ISSUE THREAD

**Issue by ooopus** [astral-sh/ty#2215]
## Summary

When using the common guard pattern:

```py
if hasattr(obj, "attr") and obj.attr:
    ...
````

ty incorrectly infers `obj.attr` as `~AlwaysFalsy`, producing a false-positive

`error[unresolved-attribute]` when accessing attributes on `obj.attr`.

This code runs correctly at runtime; the issue appears to be in ty's flow-sensitive narrowing.

## Environment

* OS: Windows 11

* Python: 3.13.9

* ty: 0.0.7 (cf82a04b5 2025-12-24)

## Command

```sh
ty check test_ty_always_falsy.py
```

(No special `pyproject.toml` settings; default config.)

## Reproduction

```py
from __future__ import annotations

class Editor:
    def __init__(self) -> None:
        self.note: Note | None = Note()

class Note:
    def __init__(self) -> None:
        self.fields: list[str] = ["field1", "field2"]

class WebView:
    pass

class EditorWebView:
    def __init__(self) -> None:
        self.editor: Editor | None = Editor()

def test(webview: WebView | EditorWebView) -> list[str] | None:
    if hasattr(webview, "editor") and webview.editor:
        reveal_type(webview)
        reveal_type(webview.editor)

        note = webview.editor.note # ty: error[unresolved-attribute]

        if note:
            return note.fields

    return None
```

## Expected Behavior

Inside:

```py
if hasattr(webview, "editor") and webview.editor:
```

`webview.editor` should be narrowed to a truthy `Editor` type, so

`webview.editor.note` should be valid.

## Actual Behavior

ty reports a false positive similar to:

```
error[unresolved-attribute]: Object of type `~AlwaysFalsy` has no attribute `note`
```

## Workarounds

Both of the following avoid the false positive:

* `isinstance()`-based narrowing

* `getattr(webview, "editor", None)` + truthiness check

### Version

ty 0.0.7 (cf82a04b5 2025-12-24)

**Comment by mtshiba** [comment]
In fact, I think this is the intended behavior, because `WebView` can be inherited and may have an attribute `editor`, so even if we can see that the parameter `webview` has an attribute `editor`, it doesn't say much about its type.

```python
from __future__ import annotations

class Editor:
    def __init__(self) -> None:
        self.note: Note | None = Note()

class Note:
    def __init__(self) -> None:
        self.fields: list[str] = ["field1", "field2"]

class WebView:
    pass

class WebViewSub(WebView):
    def __init__(self) -> None:
        self.editor = 0  # You can assign any type of object here.

class EditorWebView:
    def __init__(self) -> None:
        self.editor: Editor | None = Editor()

def test(webview: WebView | EditorWebView) -> list[str] | None:
    if hasattr(webview, "editor"):
        reveal_type(webview)  # It may be of type `WebViewSub`
        reveal_type(webview.editor)  # It may be of type `int`!

    return None

test(WebViewSub())
```

If you want to avoid this, you can use `typing.final`, and ty will behave exactly as you intended.

```python
from __future__ import annotations
from typing import final

class Editor:
    def __init__(self) -> None:
        self.note: Note | None = Note()

class Note:
    def __init__(self) -> None:
        self.fields: list[str] = ["field1", "field2"]

@final
class WebView:
    pass

class EditorWebView:
    def __init__(self) -> None:
        self.editor: Editor | None = Editor()

def test(webview: WebView | EditorWebView) -> list[str] | None:
    if hasattr(webview, "editor") and webview.editor:
        reveal_type(webview)
        reveal_type(webview.editor)

        note = webview.editor.note # OK
        reveal_type(note)  # Note | None

        if note:
            return note.fields

    return None
```

**Comment by mtshiba** [comment]
However, it's certainly not clear why narrowing failed, and diagnostic messages could be improved - for example, if the user accesses a non-existent attribute using an expression that appears inside `cond` of `if cond: ...`, it could explain the fact that narrowing failed (or is incomplete) and why.

**Comment by carljm** [comment]
I had mentioned the idea in https://github.com/astral-sh/ty/issues/1578#issuecomment-3542936632 that we could improve diagnostics for some similar cases. I will reopen this issue to track this idea. Thanks for the report!

**Comment by carljm** [comment]
Oh, I just saw #2221 -- we can use that issue instead.
